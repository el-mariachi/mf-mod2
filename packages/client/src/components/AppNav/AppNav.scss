.app-nav {
  margin-left: 20px;

  &__ {
    &toggler {
      border-width: 1px;
      border-color: rgba($color: white, $alpha: 0.25);
      border-radius: 0;
      padding: 2px 4px;

      .navbar-toggler-icon {
        background-image: url('data:image/svg+xml;utf8,<svg viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M5 6.5H19V8H5V6.5Z" fill="white"/><path d="M5 16.5H19V18H5V16.5Z" fill="white"/><path d="M5 11.5H19V13H5V11.5Z" fill="white"/></svg>');
        transform: scale(1.2);
      }
    }
    &header {
      .btn-close {
        background-image: url('data:image/svg+xml;utf8,<svg viewBox="0 0 28 28" fill="none" xmlns="http://www.w3.org/2000/svg"><g><path fill="black" d="M0,24l4,4l10-10l10,10l4-4L18,14L28,4l-4-4L14,10L4,0L0,4l10,10L0,24z"/></g></svg>');
      }
    }
  }
}
