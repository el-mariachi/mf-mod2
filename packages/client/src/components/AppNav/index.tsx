import classNames from 'classnames'
import { FC } from 'react'
import { Nav, Navbar, NavbarProps, Offcanvas } from 'react-bootstrap'
import { NavLink } from 'react-router-dom'
import { logout } from '@services/authController'
import Icon, { IconName } from '@components/Icon'
import { useAppDispatch } from '@hooks/redux_typed_hooks'
import { clearUser } from '@store/slices/user'
import { useNavigate } from 'react-router-dom'
import ROUTES from '@constants/routes'
import { AppContainerContext } from 'context'
import './AppNav.scss'

type AppNavPath = {
  title: string
  icon?: IconName
  path?: string
  callback?: () => void
  extLink?: string
}
export type AppNavProps = NavbarProps & {
  paths: AppNavPath[]
  caption?: string
}
const AppNav: FC<AppNavProps> = ({ paths, caption, className: cls }) => {
  const dispatch = useAppDispatch()
  const navigate = useNavigate()
  const navItems = paths.map(({ title, icon, path, callback, extLink }) => {
    return (
      <li key={title} className="nav-item">
        {path ? (
          <NavLink
            className={({ isActive }) =>
              classNames('nav-link d-flex align-items-center', {
                active: isActive,
              })
            }
            to={path}
            end>
            {icon ? <Icon iconName={icon} className="me-2" /> : null}
            <span>{title}</span>
          </NavLink>
        ) : (
          <Nav.Link
            onClick={(e: React.SyntheticEvent) => {
              if (callback) {
                e.preventDefault()
                callback()
              }
            }}
            className="nav-link d-flex align-items-center"
            href={extLink || '#'}>
            {title}
          </Nav.Link>
        )}
      </li>
    )
  })
  return (
    <AppContainerContext.Consumer>
      {container => (
        <Navbar className={classNames('app-nav', cls)} expand={false}>
          <Navbar.Toggle
            className="app-nav__toggler"
            aria-controls="offcanvasNavbar"
          />
          <Navbar.Offcanvas
            id="offcanvasNavbar"
            container={container}
            aria-labelledby="offcanvasNavbarLabel"
            placement="end">
            <Offcanvas.Header className="app-nav__header px-4" closeButton>
              {caption ? (
                <Offcanvas.Title
                  className="offcanvas-title minecrafted minecrafted_bold"
                  id="offcanvasNavbarLabel">
                  {caption}
                </Offcanvas.Title>
              ) : null}
            </Offcanvas.Header>
            <Offcanvas.Body className="p-4">
              <Nav className="justify-content-end flex-grow-1 pe-3" as="ul">
                {navItems}
                <li key="exit" className="nav-item">
                  <Nav.Link
                    onClick={(e: React.SyntheticEvent) => {
                      e.preventDefault()
                      logout().then(() => {
                        dispatch(clearUser())
                        navigate(ROUTES.SIGN_IN)
                      })
                    }}
                    className="nav-link d-flex align-items-center"
                    href="#">
                    <Icon iconName="XSquareFill" className="me-2" />
                    <span>Выход</span>
                  </Nav.Link>
                </li>
              </Nav>
            </Offcanvas.Body>
          </Navbar.Offcanvas>
        </Navbar>
      )}
    </AppContainerContext.Consumer>
  )
}
export default AppNav
