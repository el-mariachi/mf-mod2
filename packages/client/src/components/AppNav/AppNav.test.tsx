import AppNav from './index'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { store } from '@store/index'

describe('Testing AppNav component', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    })
  })
  it('should match snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <AppNav
          caption="One Bit Dungeon"
          paths={[
            {
              title: 'Играть!',
              path: '/',
              icon: 'Dice3Fill',
            },
            {
              title: 'Список лидеров',
              path: '/leaderboard',
              icon: 'BarChartLineFill',
            },
            {
              title: 'Игровой форум',
              path: '/forum',
              icon: 'ChatLeftTextFill',
            },
            {
              title: 'Профиль игрока',
              path: '/profile',
              icon: 'PersonSquare',
            },
          ]}
        />
      </Provider>,
      { wrapper: BrowserRouter }
    )
    expect(container).toMatchSnapshot()
  })
})
