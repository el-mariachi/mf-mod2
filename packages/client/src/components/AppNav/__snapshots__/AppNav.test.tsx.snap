// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AppNav component should match snapshot 1`] = `
<div>
  <nav
    class="app-nav navbar navbar-light"
  >
    <button
      aria-controls="offcanvasNavbar"
      aria-label="Toggle navigation"
      class="app-nav__toggler navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      aria-labelledby="offcanvasNavbarLabel"
      class="offcanvas offcanvas-end"
      id="offcanvasNavbar"
    >
      <div
        class="app-nav__header px-4 offcanvas-header"
      >
        <div
          class="offcanvas-title minecrafted minecrafted_bold offcanvas-title h5"
          id="offcanvasNavbarLabel"
        >
          One Bit Dungeon
        </div>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="p-4 offcanvas-body"
      >
        <ul
          class="justify-content-end flex-grow-1 pe-3 navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="nav-link d-flex align-items-center active"
              href="/"
            >
              <div
                class="icon me-2"
              >
                <svg
                  class="icon__img"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3 0a3 3 0 0 0-3 3v10a3 3 0 0 0 3 3h10a3 3 0 0 0 3-3V3a3 3 0 0 0-3-3H3zm2.5 4a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm8 8a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zM8 9.5a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
                  />
                </svg>
              </div>
              <span>
                Играть!
              </span>
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link d-flex align-items-center"
              href="/leaderboard"
            >
              <div
                class="icon me-2"
              >
                <svg
                  class="icon__img"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11 2a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1v12h.5a.5.5 0 0 1 0 1H.5a.5.5 0 0 1 0-1H1v-3a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1v3h1V7a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1v7h1V2z"
                  />
                </svg>
              </div>
              <span>
                Список лидеров
              </span>
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link d-flex align-items-center"
              href="/forum"
            >
              <div
                class="icon me-2"
              >
                <svg
                  class="icon__img"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 2a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2H4.414a1 1 0 0 0-.707.293L.854 15.146A.5.5 0 0 1 0 14.793V2zm3.5 1a.5.5 0 0 0 0 1h9a.5.5 0 0 0 0-1h-9zm0 2.5a.5.5 0 0 0 0 1h9a.5.5 0 0 0 0-1h-9zm0 2.5a.5.5 0 0 0 0 1h5a.5.5 0 0 0 0-1h-5z"
                  />
                </svg>
              </div>
              <span>
                Игровой форум
              </span>
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link d-flex align-items-center"
              href="/profile"
            >
              <div
                class="icon me-2"
              >
                <svg
                  class="icon__img"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11 6a3 3 0 1 1-6 0 3 3 0 0 1 6 0z"
                  />
                  <path
                    d="M2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2zm12 1a1 1 0 0 1 1 1v12a1 1 0 0 1-1 1v-1c0-1-1-4-6-4s-6 3-6 4v1a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1h12z"
                  />
                </svg>
              </div>
              <span>
                Профиль игрока
              </span>
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link d-flex align-items-center nav-link"
              data-rr-ui-event-key="#"
              href="#"
              role="button"
              tabindex="0"
            >
              <div
                class="icon me-2"
              >
                <svg
                  class="icon__img"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2zm3.354 4.646L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 1 1 .708-.708z"
                  />
                </svg>
              </div>
              <span>
                Выход
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
